internal/controller/certificates/policies/checks.go

```diff
@@ -166,6 +166,35 @@ func SecretIssuerAnnotationsNotUpToDate(input Input) (string, string, bool) {
 	return "", "", false
 }
 
+// SecretCertificateMatchesSpec checks that the current CertificateRequest contains a CSR that is
+// signed by the key stored in the Secret. A failure is often caused by the Secret being changed
+// outside of the control of cert-manager, causing the current CertificateRequest to no longer
+// match what is stored in the Secret.
+func SecretPublicKeysDiffersFromCurrentCertificateRequest(input Input) (string, string, bool) {
+	if input.CurrentRevisionRequest == nil {
+		return "", "", false
+	}
+	pk, err := pki.DecodePrivateKeyBytes(input.Secret.Data[corev1.TLSPrivateKeyKey])
+	if err != nil {
+		return InvalidKeyPair, fmt.Sprintf("Issuing certificate as Secret contains invalid private key data: %v", err), true
+	}
+
+	csr, err := pki.DecodeX509CertificateRequestBytes(input.CurrentRevisionRequest.Spec.Request)
+	if err != nil {
+		return InvalidCertificateRequest, fmt.Sprintf("Failed to decode current CertificateRequest: %v", err), true
+	}
+
+	equal, err := pki.PublicKeysEqual(csr.PublicKey, pk.Public())
+	if err != nil {
+		return InvalidCertificateRequest, fmt.Sprintf("CertificateRequest's public key is invalid: %v", err), true
+	}
```

Comment:
I don't think the CSR public key would be corrupt. it would be eg. a `ecdh.PublicKey` or `dsa.PublicKey` key (corrupt keys will give parsing errors). I don't think we can generate CSRs with such keys, so we should be fine?